File: leetcode/53-maximum-subarray.py
# https://leetcode.com/problems/maximum-subarray/ (top-interview-questions)

from typing import List

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        max_sum, curr_max = nums[0], 0
        
        for num in nums:
            if curr_max < 0: curr_max = 0
            curr_max += num
            max_sum = max(max_sum, curr_max)
        
        return max_sum

File: leetcode/53-maximum-subarray.js
// https://leetcode.com/problems/maximum-subarray/ (top-interview-questions)


/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
  
  let maxSum = Number.MIN_SAFE_INTEGER;
  let curSum = 0;

  for (const num of nums) {
    if (curSum < 0) curSum = num;
    else curSum += num;

    maxSum = Math.max(maxSum, curSum);
  }

  return maxSum;
};